library/std/src/io/buffered/bufwriter.rs

```diff
@@ -287,6 +289,77 @@ impl<W: Write> BufWriter<W> {
             Ok(()) => Ok(self.inner.take().unwrap()),
         }
     }
+
+    /// Disassembles this `BufWriter<W>`, returning the underlying writer, and any buffered but
+    /// unwritten data.
+    ///
+    /// If the underlying writer panicked, it is not known what portion of the data was written.
+    /// In this case, we return `WriterPanicked` for the buffered data (from which the buffer
+    /// contents can still be recovered).
+    ///
+    /// `into_raw_parts` makes no attempt to flush data and cannot fail.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(bufwriter_into_raw_parts)]
+    /// use std::io::{BufWriter, Write};
+    ///
+    /// let mut buffer = [0u8; 10];
+    /// let mut stream = BufWriter::new(buffer.as_mut());
+    /// write!(stream, "too much data").unwrap();
+    /// stream.flush().expect_err("it doesn't fit");
+    /// let (recovered_writer, buffered_data) = stream.into_raw_parts();
+    /// assert_eq!(recovered_writer.len(), 0);
+    /// assert_eq!(&buffered_data.unwrap(), b"ata");
+    /// ```
+    #[unstable(feature = "bufwriter_into_raw_parts", issue = "none")]
+    pub fn into_raw_parts(mut self) -> (W, Result<Vec<u8>, WriterPanicked>) {
```

Comment:
`WriterPanicked` feels weird, I wonder if it is able to just return `self.panicked`?

Reply:
Just returning `panicked` seems rather uncooked.  See my commit message: it would be easy for someone to write code which just threw `panicked` away and then used the buffer anyway.  This way the buffer is type-entangled with the error condition, which seems right to me.

In other words, I think this arrangement is probably the *best* arrangement, but the question is just whether this additional stuff is carrying its weight in terms of API complexity.  I think it is, mostly because the way rustdoc etc. format these things means you don't have to read it most of the time.